Widgets/Widgets/TopReadWidget.swift

```diff
@@ -0,0 +1,371 @@
+import SwiftUI
+import WidgetKit
+import WMF
+
+// MARK: - Widget
+
+struct TopReadWidget: Widget {
+	private let kind: String = WidgetController.SupportedWidget.topRead.identifier
+
+	public var body: some WidgetConfiguration {
+		StaticConfiguration(kind: kind, provider: TopReadProvider(), content: { entry in
+			TopReadView(entry: entry)
+		})
+		.configurationDisplayName(LocalizedStrings.topReadWidgetTitle)
+		.description(LocalizedStrings.topReadWidgetDescription)
+		.supportedFamilies([.systemSmall, .systemMedium, .systemLarge])
+	}
+}
+
+// MARK: - Data
+
+final class TopReadData {
+
+	// MARK: Properties
+
+	static let shared = TopReadData()
+
+	let maximumRankedArticles = 4
+
+	var placeholder: TopReadEntry {
+		return TopReadEntry(date: Date())
+	}
+
+	// MARK: Public
+
+	private var dataStore: MWKDataStore {
+		MWKDataStore.shared()
+	}
+
+	func fetchLatestAvailableTopRead(_ completion: @escaping (TopReadEntry) -> Void) {
+		guard
+			let appLanguage = dataStore.languageLinkController.appLanguage,
+			let topRead = dataStore.viewContext.group(of: .topRead, for: Date(), siteURL: appLanguage.siteURL()),
```

Comment:
This looks like it does what I was thinking of with POTD - fetching the latest content group for today and the primary language. I was also going to suggest we trigger a widget reload after the user changes their primary language in the app. Though on further thought hooking into Explore Feed's content reloading might cover it since that updates with a primary language change as well.

Reply:
Just a documentation comment, confirmed that updating languages hits `WMFAppViewControllers.m's `preferredLanguagesDidChange` method, which eventually calls the `updateFeedSourcesWithDate` method where the `reloadAllWidgets` call lives. So no need for separate reload call on language change.